File: scripts/lib/fill_search_options.py
"""fillSearch 选项 — 与 fill-search-options.mjs 对齐。"""
from __future__ import annotations

from typing import Any


def _as_string_list(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, str):
        s = value.strip()
        return [s] if s else []
    if isinstance(value, list):
        return [str(x).strip() for x in value if str(x).strip()]
    return []


def _merge_unique(*lists: list[str]) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for lst in lists:
        for item in lst:
            key = item.lower()
            if key in seen:
                continue
            seen.add(key)
            out.append(item)
    return out


def fill_search_payload_from_arg(hints_or_opts: Any) -> dict[str, Any]:
    if hints_or_opts is None:
        return {}
    if isinstance(hints_or_opts, str) or isinstance(hints_or_opts, list):
        labels = _as_string_list(hints_or_opts)
        if not labels:
            return {}
        return {"uiHeuristics": {"searchEntryLabels": labels, "searchInputLabels": labels}, "hints": labels}

    if not isinstance(hints_or_opts, dict):
        return {}

    entry_hints = _as_string_list(hints_or_opts.get("entryHints"))
    input_hints = _as_string_list(hints_or_opts.get("inputHints"))
    legacy = _as_string_list(hints_or_opts.get("hints"))
    ui: dict[str, Any] = {}
    merged_entry = _merge_unique(entry_hints, legacy)
    merged_input = _merge_unique(input_hints, legacy)
    if merged_entry:
        ui["searchEntryLabels"] = merged_entry
    if merged_input:
        ui["searchInputLabels"] = merged_input

    out: dict[str, Any] = {}
    if ui:
        out["uiHeuristics"] = ui
    if entry_hints:
        out["entryHints"] = entry_hints
    if input_hints:
        out["inputHints"] = input_hints
    if legacy:
        out["hints"] = legacy
    if hints_or_opts.get("strict") is True:
        out["strict"] = True
    settle = hints_or_opts.get("settleMs")
    if isinstance(settle, (int, float)):
        out["settleMs"] = int(settle)
    if hints_or_opts.get("skipRedundantDump") is True:
        out["skipRedundantDump"] = True
    return out

File: scripts/lib/fill-search-options.mjs
/** fillSearch 选项解析 — 与 packages/driver-rpc/src/fill-search-options.ts 同步 — run: npm run sync:scripts-lib */

function asStringList(v) {
    if (v == null)
        return [];
    if (typeof v === "string")
        return v.trim() ? [v.trim()] : [];
    if (Array.isArray(v)) {
        return v.map((x) => String(x).trim()).filter(Boolean);
    }
    return [];
}
function mergeUnique(...lists) {
    const out = [];
    const seen = new Set();
    for (const list of lists) {
        for (const item of list) {
            const key = item.toLowerCase();
            if (seen.has(key))
                continue;
            seen.add(key);
            out.push(item);
        }
    }
    return out;
}
/** 从 recipe/custom payload 解析 fillSearch 选项（P1） */
export function parseFillSearchPayload(payload) {
    const p = payload ?? {};
    const nested = typeof p.fillSearch === "object" && p.fillSearch !== null
        ? p.fillSearch
        : {};
    const ui = (p.uiHeuristics ?? nested.uiHeuristics);
    const legacyHints = asStringList(p.hints ?? nested.hints);
    const entryHints = mergeUnique(asStringList(p.entryHints ?? nested.entryHints), asStringList(ui?.searchEntryLabels), legacyHints);
    const inputHints = mergeUnique(asStringList(p.inputHints ?? nested.inputHints), asStringList(ui?.searchInputLabels), legacyHints);
    const heuristics = entryHints.length || inputHints.length || ui
        ? {
            ...ui,
            ...(entryHints.length ? { searchEntryLabels: entryHints } : {}),
            ...(inputHints.length ? { searchInputLabels: inputHints } : {})
        }
        : ui;
    const strict = p.strict === true || nested.strict === true;
    const settleMs = typeof p.settleMs === "number"
        ? p.settleMs
        : typeof nested.settleMs === "number"
            ? nested.settleMs
            : undefined;
    return {
        heuristics,
        entryHints,
        inputHints,
        strict,
        recipeOptions: {
            settleMs,
            skipRedundantDump: p.skipRedundantDump === true || nested.skipRedundantDump === true,
            payload: p
        }
    };
}
/** 脚本层 hints 参数 → recipe payload 字段 */
export function fillSearchPayloadFromArg(hintsOrOpts) {
    if (hintsOrOpts == null)
        return {};
    if (typeof hintsOrOpts === "string" || Array.isArray(hintsOrOpts)) {
        const list = asStringList(hintsOrOpts);
        return list.length
            ? { uiHeuristics: { searchEntryLabels: list, searchInputLabels: list }, hints: list }
            : {};
    }
    const entryHints = asStringList(hintsOrOpts.entryHints);
    const inputHints = asStringList(hintsOrOpts.inputHints);
    const legacy = asStringList(hintsOrOpts.hints);
    const ui = {
        ...(entryHints.length || legacy.length
            ? { searchEntryLabels: mergeUnique(entryHints, legacy) }
            : {}),
        ...(inputHints.length || legacy.length
            ? { searchInputLabels: mergeUnique(inputHints, legacy) }
            : {})
    };
    const out = {};
    if (Object.keys(ui).length)
        out.uiHeuristics = ui;
    if (entryHints.length)
        out.entryHints = entryHints;
    if (inputHints.length)
        out.inputHints = inputHints;
    if (legacy.length)
        out.hints = legacy;
    if (hintsOrOpts.strict === true)
        out.strict = true;
    if (typeof hintsOrOpts.settleMs === "number")
        out.settleMs = hintsOrOpts.settleMs;
    if (hintsOrOpts.skipRedundantDump === true)
        out.skipRedundantDump = true;
    return out;
}
